bundles/org.openhab.binding.bmwconnecteddrive/src/main/java/org/openhab/binding/bmwconnecteddrive/internal/handler/ConnectedDriveProxy.java

```diff
@@ -0,0 +1,315 @@
+     * So if token refresh fails the corresponding requests will also fail and update the
+     * Thing status accordingly.
+     *
+     * @return token
+     */
+    public Token getToken() {
+        if (token.isExpired() || !token.isValid()) {
+            updateToken();
+        }
+        return token;
+    }
+
+    /**
+     * Authorize at BMW Connected Drive Portal and get Token
+     *
+     * @return
+     */
+    private synchronized void updateToken() {
+        Request req = authHttpClient.POST(legacyAuthUri);
+
+        req.header(HttpHeader.CONTENT_TYPE, CONTENT_TYPE_URL_ENCODED);
+        req.header(HttpHeader.CONNECTION, KEEP_ALIVE);
+        req.header(HttpHeader.HOST, BimmerConstants.SERVER_MAP.get(configuration.region));
+        req.header(HttpHeader.AUTHORIZATION, BimmerConstants.AUTHORIZATION_VALUE_MAP.get(configuration.region));
+        req.header(CREDENTIALS, BimmerConstants.CREDENTIAL_VALUES);
+        req.header(HttpHeader.REFERER, BimmerConstants.REFERER_URL);
+
+        MultiMap<String> dataMap = new MultiMap<String>();
+        dataMap.add("grant_type", "password");
+        dataMap.add(SCOPE, BimmerConstants.SCOPE_VALUES);
+        dataMap.add(USERNAME, configuration.userName);
+        dataMap.add(PASSWORD, configuration.password);
+        String urlEncodedData = UrlEncoded.encode(dataMap, Charset.defaultCharset(), false);
+        req.header(CONTENT_LENGTH, Integer.toString(urlEncodedData.length()));
+        req.content(new StringContentProvider(urlEncodedData));
+        try {
+            ContentResponse contentResponse = req.timeout(HTTP_TIMEOUT_SEC, TimeUnit.SECONDS).send();
+            // Status needs to be 302 - Response is stored in Header
+            if (contentResponse.getStatus() == 302) {
+                HttpFields fields = contentResponse.getHeaders();
+                HttpField field = fields.getField(HttpHeader.LOCATION);
+                tokenFromUrl(field.getValue());
+            } else if (contentResponse.getStatus() == 200) {
+                AuthResponse authResponse = Converter.getGson().fromJson(contentResponse.getContentAsString(),
+                        AuthResponse.class);
+                if (authResponse != null) {
+                    token.setToken(authResponse.accessToken);
+                    token.setType(authResponse.tokenType);
+                    token.setExpiration(authResponse.expiresIn);
+                }
+            } else {
+                logger.debug("Authorization status {} reason {}", contentResponse.getStatus(),
+                        contentResponse.getReason());
+            }
+        } catch (InterruptedException | ExecutionException | TimeoutException e) {
+            logger.debug("Authorization exception: {}", e.getMessage());
+            StackTraceElement[] trace = e.getStackTrace();
+            for (int i = 0; i < trace.length; i++) {
+                logger.info("{}", trace[i]);
+            }
```

Comment:
Why are you manually writing the trace when you can include the entire exception as part of the logged message?
```suggestion
            logger.debug("Authorization exception: {}", e.getMessage(), e);
```

Reply:
fixed